rosidl_typesupport_opensplice_cpp/resource/visibility_control.h.in

```diff
@@ -1,8 +1,7 @@
 // generated from rosidl_typesupport_opensplice_cpp/resource/visibility_control.h.in
 // generated code does not contain a copyright notice
 
-#ifndef @PROJECT_NAME_UPPER@__MSG__ROSIDL_TYPESUPPORT_OPENSPLICE_CPP__VISIBILITY_CONTROL_H_
-#define @PROJECT_NAME_UPPER@__MSG__ROSIDL_TYPESUPPORT_OPENSPLICE_CPP__VISIBILITY_CONTROL_H_
+// NOLINT(build/header_guard)
```

Comment:
Why shouldn't this header file have any guard?

Reply:
> update visibility template, OpenSplice generated code (idlpp output)
undefs visibility defines. When combining generated type from multiple
input files this leads to compile erros as the required visibility
define is undefined by the first type.

This generated file is included by the idlpp generated code. The idlpp generated code undefines what is defined in this file. When a C/C++ source file includes multiple idlpp generated header files which all use this generated file the first one will set the guard set the defines and unset the defines. The following files will run into the guard and thus not set the defines and have their visibility wrong which causes a build error. Since it's possible to redefine X as long as it's syntactically identical a guard is not needed.
https://msdn.microsoft.com/en-us/library/teas0593.asps

Reply:
> The idlpp generated code undefines what is defined in this file.

Sorry, that I am not understanding this yet. How does generated code undefine what is defined in this file? The `idlpp` doesn't know about the types defined in this file.

Can you point to a file / template or paste snippet of the generated code describe it?

> The following files will run into the guard and thus not set the defines and have their visibility wrong which causes a build error.

In which case does that happen? The latest builds seemed fine and passed and I am sure we do have message with nested structures.

Reply:
idlpp is called with the -P option.

    -P dll_macro_name[,<header-file>]
      Only applicable to C and C++. Sets export macro that will be
      prefixed to all functions in the generated code. This allows
      creating DLL's from generated code and controlling symbol visibility.
      Optionally a header file can be given that will be included
      in each generated file.

where dll_macro_name=ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_PUBLIC_@PROJECT_NAME@ and header_file=@PROJECT_NAME@...visibility_control.h. When using the GoalID_.idl from the actionlib_msgs project the idlpp output looks like:

// actionlib_msgs/rosidl_typesupport_opensplice_cpp/actionlib_msgs/msg/dds_opensplice/GoalID_.h
#ifndef _GOALID__H_
#define _GOALID__H_

#include "sacpp_mapping.h"
#include "dds_dcps.h"
#include "actionlib_msgs/msg/rosidl_typesupport_opensplice_cpp__visibility_control.h"

<<CODE REMOVED>>
    struct ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_PUBLIC_actionlib_msgs GoalID_
    {
<<CODE REMOVED>>

#undef ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_PUBLIC_actionlib_msgs

#endif

so the actionlib_msgs/msg/rosidl_typesupport_opensplice_cpp__visibility_control.h has a #define for ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_PUBLIC_actionlib_msgs and in the above displayed code snippet you can see the #undef.
The actionlib_msgs project also has a GoalStatusArray_.idl file which uses the same -P option for idlpp and thus does the same include "actionlib_msgs/msg/rosidl_typesupport_opensplice_cpp__visibility_control.h" and requires ROSIDL_TYPESUPPORT_OPENSPLICE_CPP_PUBLIC_actionlib_msgs which is now on longer defined.

Reply:
I am not sure if this is the best way. In order to be able to test this I guess we need an up-to-date version of OpenSplice for Windows which we can install on our Jenkins nodes (OS X would be good to). Currently we only have our self-made Debian package for the Linux nodes. Can you provide us with these so that we can CI on this (and I can look at it a bit closer)?

Reply:
I agree, this is not the best way. Unfortunately I cannot change the idlpp generated code because of backwards incompatibility. We could remove the #undef post-generation.

Will look into getting you a Windows build on Monday when back in the office (macOS going to be a bit harder since it's not officially supported yet but we have some machines running it).

Reply:
I installed the shared package in one of our Windows nodes and triggered a build with this branch: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows_opensplice67&build=3)](http://ci.ros2.org/job/ci_windows_opensplice67/3/)

Reply:
I installed the shared package in one of our Windows nodes and triggered a build with this branch: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows_opensplice67&build=3)](http://ci.ros2.org/job/ci_windows_opensplice67/3/)

**Update**: the build failed with: `cppgen : error : unable to open file: JointState_Dcps.h`

Reply:
I've analyzed the build issue and my preliminary conclusion is that the generation of the of "type support for PrismTech OpenSplice" is done multiple times. I've also been able to verify this on Linux when building with -j16. My cmake knowledge is not as good as I want it to, so I'm not able to fix this issue atm. A work around for now would be to not build parallel.

Reply:
I started a no-parallel build: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows_opensplice67&build=4)](http://ci.ros2.org/job/ci_windows_opensplice67/4/)

I will look if I can reproduce the problem. Surprisingly this hasn't happened on Linux before.

Reply:
And another build (the previous one failed with permission problems - not sure what happened there on the build node): [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows_opensplice67&build=5)](http://ci.ros2.org/job/ci_windows_opensplice67/5/)

Reply:
I am not sure but maybe ros2/rmw_opensplice#195 will help with the problem with the generator we saw in the other build.

Have you made any progress on the new compiler warnings? Or do you have a newer Windows package for me to try?